arch/x86/core/intel64/irq.c

```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright (c) 2019 Intel Corporation
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <kernel.h>
+#include <arch/cpu.h>
+#include <kernel_arch_data.h>
+#include <drivers/interrupt_controller/sysapic.h>
+#include <irq.h>
+
+unsigned char _irq_to_interrupt_vector[CONFIG_MAX_IRQ_LINES];
+
+/*
+ * The low-level interrupt code consults these arrays to dispatch IRQs, so
+ * so be sure to keep locore.S up to date with any changes. Note the indices:
+ * use (vector - IV_IRQS), since exception vectors do not appear here.
+ *
+ * Entries which are NULL in x86_irq_funcs[] correspond to unassigned vectors.
+ * The locore IRQ handler should (read: doesn't currently) raise an exception
+ * rather than attempt to dispatch to a NULL x86_irq_func[]. FIXME.
+ */
+
+#define NR_IRQ_VECTORS (IV_NR_VECTORS - IV_IRQS)  /* # vectors free for IRQs */
+
+void (*x86_irq_funcs[NR_IRQ_VECTORS])(void *);
+void *x86_irq_args[NR_IRQ_VECTORS];
+
+/*
+ * Interrupt stack.
+ */
+
+char __aligned(STACK_ALIGN) ist[CONFIG_ISR_DEPTH][CONFIG_ISR_STACK_SIZE];
+
+/*
+ * Find a free IRQ vector at the specified priority, or return -1 if none left.
+ */
+
+static int allocate_vector(unsigned int priority)
+{
+	const int VECTORS_PER_PRIORITY = 16;
+	const int MAX_PRIORITY = 13;
+
+	int vector;
+	int i;
+
+	if (priority >= MAX_PRIORITY) {
+		priority = MAX_PRIORITY;
+	}
+
+	vector = (priority * VECTORS_PER_PRIORITY) + IV_IRQS;
+
+	for (i = 0; i < VECTORS_PER_PRIORITY; ++i, ++vector) {
+#ifdef CONFIG_IRQ_OFFLOAD
+		if (vector == CONFIG_IRQ_OFFLOAD_VECTOR) {
+			continue;
+		}
+#endif
+		if (x86_irq_funcs[vector - IV_IRQS] == NULL) {
+			return vector;
+		}
+	}
+
+	return -1;
+}
+
+/*
+ * N.B.: the API docs don't say anything about returning error values, but
+ * this function returns -1 if a vector at the specific priority can't be
+ * allocated. Whether it should simply __ASSERT instead is up for debate.
+ */
+
+int z_arch_irq_connect_dynamic(unsigned int irq, unsigned int priority,
+		void (*func)(void *arg), void *arg, u32_t flags)
+{
+	u32_t key;
+	int vector;
+
+	__ASSERT(irq <= CONFIG_MAX_IRQ_LINES, "IRQ %u out of range", irq);
+
+	key = irq_lock();
+
+	vector = allocate_vector(priority);
+	if (vector >= 0) {
+		_irq_to_interrupt_vector[irq] = vector;
+		z_irq_controller_irq_config(vector, irq, flags);
+		x86_irq_funcs[vector - IV_IRQS] = func;
+		x86_irq_args[vector - IV_IRQS] = arg;
+	}
+
+	irq_unlock(key);
+	return vector;
+}
+
+#ifdef CONFIG_IRQ_OFFLOAD
+#include <irq_offload.h>
+
+void irq_offload(irq_offload_routine_t routine, void *parameter)
+{
+	u32_t key;
+
+	key = irq_lock();
```

Comment:
The existing 32-bit x86 code uses `irq_lock()` here, which is why I've kept it -- with SMP this expands to `z_smp_global_lock()`, which I assume preserves the UP semantics on SMP systems.

The truth is that I'm not really sure we care about serialization here, since `irq_offload()` is only for testing, right? Do we even care about making it thread-safe?  If we do, is a spinlock appropriate?  I would think this would mean the "offloaded" function can't do anything that might cause a context switch ... (though I'm uncertain of that restriction - perhaps @andyross can clarify, I can't find any documentation on this).

I'd prefer to leave this as-is and, if a change is necessary, to change both 32- and 64-bit code simultaneously (later).  Seems to work OK for the moment in on our uniprocessor targets.

Reply:
Right, but UP semantics of an irq_lock() are "nothing else in the system will run", which is exactly what we don't want with SMP and the reason the API needed to be replaced.  The goal is to have zero irq_lock() calls in Zephyr code long term, though obviously we can drag our feet for stuff like drivers on non-SMP platforms.  Here we really want to be forward-looking.

The lock, though, does need to be released before the INT instruction (or be treated as so by its handler).  We want an irq_offload() function to look like an interrupt, not an exception.

But broadly I think you're right about irq_offload.  It's badly named, poorly documented and inconsistently implemented.   Other platforms aren't very consistent about synchronization.

Reply:
> Other platforms aren't very consistent about synchronization.

I'm going to leave this alone for now.  The `irq_lock()` functions properly here to serialize the use of the global variables when UP (albeit in a not-quite-obvious fashion) and, as you say, the question of properly synchronizing this under other circumstances has not been decided.